### PR TITLE
QA-14485 Add filter to handle multiple consecutive directives

### PR DESCRIPTION
**Pull request by AKarmanov**
## JIRA

https://jira.jahia.org/browse/QA-14485

## Description

Basic filter to prevent multiple consecutive directives which can be used for attack

### CODE REVIEW

**Comment by AKarmanov** [comment]
I need to improve regex the one used is too permissive

File: graphql-dxm-provider/src/main/java/org/jahia/modules/graphql/provider/dxm/security/MultiReadRequestWrapper.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package org.jahia.modules.graphql.provider.dxm.security;
```
Comment (bpapez) [inline graphql-dxm-provider/src/main/java/org/jahia/modules/graphql/provider/dxm/security/MultiReadRequestWrapper.java:1]: Shouldn't we add these two classes to a separate module? @fpral says in the linked ticket

> we need to find a mitigation way for the following Jahia versions:
> 
>     Jahia 7.3.9 and 7.3.10
>     Jahia 8.0.3
>     Jahia 8.1.1 and 8.1.2
> 

Putting it into graphql-dxm-provider, we would need to create minor versions for all possible or used graphql-dxm-provider version on these mentioned Jahia versions. Having the filter in an own module we could more easily give it to customers, who found and are concerned by the known DoS vulnerability or even decide to pro-actively communicate that all customers should install it.

File: graphql-dxm-provider/src/main/java/org/jahia/modules/graphql/provider/dxm/security/QueryFilter.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import java.io.*;
+import java.util.regex.Pattern;
+
+import static java.util.stream.Collectors.joining;
+
+@Component(service = {Filter.class}, property = {"pattern=/graphql"}, immediate = true)
+public class QueryFilter implements Filter {
+
+    // Detect strings such as @lala@lala@lala... and with space(s) in between @lala @lala
+    private Pattern regex = Pattern.compile("@[^ ]+@|@[^ ]+[ ]+@");
```
Comment (bpapez) [inline graphql-dxm-provider/src/main/java/org/jahia/modules/graphql/provider/dxm/security/QueryFilter.java:16]: Could there be valid queries, where @ is used more than once in a string? As the performance issue only happens with many thousands of tokens, could we perhaps use regular expression, which allows a certain number of tokens to not filter out valid queries?

**Review by bpapez — CHANGES_REQUESTED** [review]
we need to find a mitigation way for the following Jahia versions:

    Jahia 7.3.9 and 7.3.10
    Jahia 8.0.3
    Jahia 8.1.1 and 8.1.2

File: graphql-dxm-provider/src/main/java/org/jahia/modules/graphql/provider/dxm/security/MultiReadRequestWrapper.java (hunk 1 above)
Comment (AKarmanov) [reply graphql-dxm-provider/src/main/java/org/jahia/modules/graphql/provider/dxm/security/MultiReadRequestWrapper.java:1]: Sure, I can try to move it out into a separate module

File: graphql-dxm-provider/src/main/java/org/jahia/modules/graphql/provider/dxm/security/QueryFilter.java (hunk 2 above)
Comment (AKarmanov) [reply graphql-dxm-provider/src/main/java/org/jahia/modules/graphql/provider/dxm/security/QueryFilter.java:16]: I did some investigating and you're right we can have multiple directives, I'll adjust regex. I think if we aim for say 10 we should be ok or do you have another number in mind?

**Pull request closed without merging**